### ISSUE TITLE
Crypto roadmap

### ISSUE THREAD

**Issue by enkore** [borgbackup/borg#1044]
Coordination and high-level discussion ticket
  - [ ] "integrate AEAD-style crypto code" (this also adds AES-GCM crypto code, which would be unused at first). At first, only the legacy-compatible mode of the code would be used.
  - [ ] Implement _key-agility_ and _algorithm-agility_ via DEKs
    - [x] Discussion in #1031 and linked spec draft
    - [x] Low-level AEAD-style abstractions in #1034 (with AES-CTR compatible to older borg repos and AES-GCM)
  - [x] require OpenSSL >= 1.1.1 (via #1034)
    - [x] AES-256-OCB? 25-50 % faster than AES-GCM (with _and_ without AES-NI). Limited data/key is not a problem since we'll have key agility at this point (also, GCM has the exact same problem and is pretty much the same code-wise; just share the code!) [License: Borg is eligible, no explicit licensing needed, since we're OSI] - Ie. OCB is pretty much the same thing for us as GCM, just faster.
    - [x] Chacha20-Poly1305 - is great because it's AEAD crypto that's faster than any AES-AEAD mode in software.

Note: edited / updated 2022-02 by TW

**Comment by ThomasWaldmann** [comment]
First item has to be "integrate AEAD-style crypto code" (this also adds AES-GCM crypto code, which would be unused at first).

**Comment by FelixSchwarz** [comment]
Not sure about the openssl version but I'd like to mention that RHEL 7 uses a (highly patched) openssl 1.0.1e. Of course I can stay with borg 1.0 but if possible I'd like to use the latest version of borg on RHEL 7 for 1-2 more years (maybe until RHEL 8 got published?).

**Comment by ThomasWaldmann** [comment]
@FelixSchwarz sounds like rhel7 could at least be used for openssl based aes-gcm mode.

Also, libsodium seems to be in EPEL7, so I guess this would be easier for you than openssl 1.1 (as there is no openssl at all in EPEL)?

**Comment by FelixSchwarz** [comment]
"EPEL" contains only extra packages for RHEL/CentOS. The "base os" (including openssl) is of course part of Red Hat's "RHEL" product (with their "10 years of support" offering if you pay for a contract). Versions in RHEL are updated *very* rarely (instead they do a lot of patching - similar to Debian stable, just much longer). Packages in EPEL must not conflict with RHEL's packages.

Officially EPEL has the same policy but as it is mostly volunteer-driven, version upgrades are easier. Which means RHEL 7 will likely never upgrade openssl. libsodium is in EPEL so as long as its version is not too old I'm fine with that.

**Comment by enkore** [comment]
From #1030 

> https://news.ycombinator.com/item?id=11634215
>
>https://github.com/fpgaminer/preserve/blob/master/crypto-spec.md

**Comment by enkore** [comment]
> https://github.com/fpgaminer/preserve/blob/master/crypto-spec.md

Similar idea popped up in a discussion with @textshell

chunk_id := HMAC(id_key, chunk)

chunk_key := HKDF(key_secret, chunk_id)

enc := header || E(chunk_key, chunk, iv=0)

**Comment by enkore** [comment]
https://github.com/borgbackup/borg/issues/1044#issuecomment-233490900 : I don't quite see the point of per-chunk keys. I derived a similar (and somewhat obvious) scheme for deterministic calculation of random IVs, and there I see the point, not so much for entire keys. That is, assuming use of session keys already.

**Comment by ThomasWaldmann** [comment]
OpenSSL 1.1 adoption in borg might be more or less blocked by #1550.

**Comment by enkore** [comment]
Discussion/analysis of keyblob (keyfile, repokey) saving/encryption: #1579

**Comment by enkore** [comment]
Python 3.6 will include BLAKE2 with a comprehensive API: https://docs.python.org/3.6/library/hashlib-blake2.html

Also new are SHA-3 and the scrypt KDF.

I've been lately thinking about adding the proposed BLAKE2 ID hash modes to 1.1. We could use the BLAKE2 reference implementation at first, which is written in portable C. (CPython 3.6 does it that way).

**Comment by ThomasWaldmann** [comment]
blake2b would be nice, but:
- we don't have much time left for 1.1 (at least not if we do not want to miss next debian stable deadline)
- how to differentiate between the different id hashes? just for new repos and a repo config key for id hash algorithm?

**Comment by enkore** [comment]
Yes. I've read #1562, but I don't think the mechanism is ripe for 1.1. And given how slowly the field (of hash functions) evolves I'm not sure a more flexible mechanism is needed -- SHA-2 was introduced in ~2000ish, BLAKE in ~2013ish.

**Comment by ThomasWaldmann** [comment]
Ok, new ticket for blake2b 256bit id hash for 1.1 milestone?

**Comment by enkore** [comment]
Crypto for 1.1 is now final, I guess.

**Comment by ThomasWaldmann** [comment]
I'ld like to put bounties on some stuff, but i need clear, separate tickets for that.

**Comment by enkore** [comment]
Here's a little comparison. The test set is a checked out git of the Linux kernel. The repository resides on /tmp (my SSD is too full). The profiler is cProfile, the visualizer is KCachegrind, the data converter is pyprof2calltree.

This is 1.0.9rc1+n. It took 42 seconds wall clock time. repokey mode.

![1.0.9](http://i.imgur.com/VtrvuEe.png)

This is 1.1.0x. It took 43 seconds wall clock time. repokey mode.

![1.1.0x](http://i.imgur.com/IFUThUB.png)

This is 1.1.0x in repokey-blake2b mode. It took 30 seconds wall clock time.

![blake2](http://i.imgur.com/s8yLuwO.png)

**Comment by ThomasWaldmann** [comment]
I'ld like to put bounties on sub-tasks of the borg 1.2 crypto goals, but to do that, I need individual tickets per subtask. We also need these tickets for the 1.2 milestone. A new (more cleaned up than this one) overview ticket or project would be also great.

**Comment by rugk** [comment]
Just because I am curious:
> their [= libsodium's] AES-GCM implementation is not an option.

Why?

**Comment by enkore** [comment]
Because it requires AES-NI (or NEON iirc), which is pretty much the only sane way to do GCM anyway from a crypto perspective, but it's bad from a usability perspective. "Sorry, you can't access this repository on this computer because your CPU does not support ASKLDJ" does not inspire trust in data accessibility.

**Comment by enkore** [comment]
OpenSSL 1.1 is in Debian 9 (Stretch): https://packages.debian.org/stretch/libssl1.1

Also Arch Linux is now on 1.1 exclusively since a while. (#2505)

**Comment by ArchangeGabriel** [comment]
> Also Arch Linux is now on 1.1 exclusively since a while. (#2505)

Not exactly true: https://www.archlinux.org/packages/?q=openssl-1.0. But yes, OpenSSL 1.1 is the new default since ~1 month.

**Comment by bpereto** [comment]
OpenSSL 1.1 will be in Fedora 26. The release is scheduled by end of june 2017.   
Centos 7/RHEL 7 will stay on OpenSSL 1.0.1

**Comment by enkore** [comment]
1.0 should be supported with critical fixes [1] until it drops out of our distribution yardstick = Debian Jessie. Which means ~2018 (or LTS ~2020, I dunno). After that we got 1.1 which also still works with OpenSSL 1.0. Given the availability of the binaries it could be okay for 1.2 to require OpenSSL 1.1.

Note that development of 1.2 hasn't even started yet, so it's not clear when it's done. Although 1.2 actually has a [defined set of goals](https://github.com/borgbackup/borg/wiki/Borg-1.2), [unlike 1.1](https://github.com/borgbackup/borg/wiki/Borg-1.1), so it's much easier for us to know when we're "done".

[1] https://borgbackup.readthedocs.io/en/latest/development.html#branching-model

**Comment by rugk** [comment]
Throwing in some info/opinion about AES-GCM… https://twitter.com/tqbf/status/760907360319660032

(BTW, but still better than Encrypt-and-MAC, https://github.com/borgbackup/borg/issues/747…)

**Comment by gthieleb** [comment]
Can one summarize the showstopper for public key encryption (GPG like) or is this discussed in another issue? Reason: I don't want to provide a passphrase for headless backup jobs.

**Comment by anarcat** [comment]
@gthieleb you should look at https://github.com/borgbackup/borg/issues/672#issuecomment-303695078

**Comment by enkore** [comment]
When changing something about the keyfile, see whether the keyfile encryption can be fixed at the same time.

> I wrote about key encrytion [here](http://borgbackup.readthedocs.io/en/stable/internals/security.html#offline-key-security); to summarize, there are ~~two~~ three issues with it:
>
> 1. It uses Encrypt-**and**-MAC in one of the few constellations not being outright broken (CTR mode, no padding and HMAC).
> 2. It reuses the same key for both AES and HMAC. Again not obviously broken in this specific constellation, but very poor practice.
> 3. It uses PBKDF2, and it uses comparatively weak parameters. (#747, another issue number I know by heart ;)

From https://github.com/borgbackup/borg/issues/3585#issuecomment-362617799

**Comment by KizzyCode** [comment]
What about considering a SIV scheme for future AE(AD) constructions? This would immediately resolve any risk of nonce reuse without the need to maintain any kind of nonce state etc.

Because the current situation is quite unsatisfying. While it is pretty easy to use only one device per repo, the advise
> You can avoid this by not continuing to write to the repository after you have lost clientside borg information.
is simply not practical unless you use borg for absolute emergency recoveries only.

**Comment by ThomasWaldmann** [comment]
aes-gcm-siv looks pretty nice and quite fast and the spec reads like made for our use case - thanks for the pointer!

but that is rather new and not present in openssl (yet?).

**Comment by KizzyCode** [comment]
> but that is rather new and not present in openssl (yet?).

I'm no OpenSSL expert, but according to the [changelog](https://www.openssl.org/news/changelog.html) SIV was implemented (search for SIV). The API should be something like `EVP_aead_aes_128_gcm_siv` or `EVP_aead_aes_256_gcm_siv`.

On the age of SIV: There are multiple flavors of SIV. The oldest construction is the classic AES-SIV (which uses AES-CMAC as MAC) was [published 2007](https://web.cs.ucdavis.edu/~rogaway/papers/siv.pdf). Then there is AES-GCM-SIV which is [much newer](https://eprint.iacr.org/2017/168.pdf). Or AES-SIV-PMAC as implemented by [miscreant](https://github.com/miscreant/miscreant.py).

However since SIV is a generic construction, you're not even limited to AES. E.g. jedisct1 (the author of libsodium) has also created a trivial [XChaCha20-Blake2b-SIV implementation](https://libsodium.gitbook.io/doc/quickstart#is-encryption-without-nonces-possible) ([repo](https://github.com/jedisct1/libsodium-xchacha20-siv)), which could be an option if you decide to opt-in for libsodium.

**Comment by rugk** [comment]
> On a platform where counters cannot be maintained, and where no trusted source of randomness exist, deterministic encryption can be used as a last resort.

…does not sound very confident though and rather like "if you cannot use anything else, do this, but we do not recommend it"…

**Comment by KizzyCode** [comment]
SIV guarantees the following:
> But SIV has strong privacy and authenticity properties even if this nonce should somehow get reused: even then, new (header, ciphertext) pairs cannot be forged, and privacy will be compromised only to the extent that an adversary can detect repetitions in (header, plaintext) pairs.

libsodium and it's documentation are designed for people who don't know much about encryption – and therefore this warning is absolutely necessary: even the simple knowledge of data repetition can be enough to break some poorly designed protocols.

However whether is a problem or not completely depends on the protocol/security model – in some cases it may even be desirable to have a deterministic bidirectional plaintext↔︎ciphertext mapping (e.g. the file names in an encrypting filesystem).

And when it comes to Borg this also should be completely fine, because a malicious server also can easily detect plaintext repetitions (and much more like access patterns etc.).

**Comment by ThomasWaldmann** [comment]
@KizzyCode i can't find aes-gcm-siv in openssl, can you point me to the declaration or implementation via a URL?

**Comment by ThomasWaldmann** [comment]
@rugk i read the SIV RFCs and they make clear that using SIV with a rarely-repeating "nonce" is of course much better than using SIV with a fixed "nonce".

that sounds pretty perfect for what we need:
- with out current counter management, in case of "accidents", like "all trusted bookkeeping lost"
- with a potential future random based nonce, in case of a repeating nonce case (which might be unlikely, depending on quality of rng and size of nonces, but possible)

i also liked that siv is able to take a vector of variable length strings and can merge them together for the AD part of AEAD.

the older implementations seem to be a bit slower than aes-gcm-siv, so that's why the latter might be nice for us.

**Comment by KizzyCode** [comment]
@ThomasWaldmann 
Oh, my bad… I read about SIV in the [OpenSSL's changelog](https://www.openssl.org/news/changelog.html) and did a quick search without noticing that all results were from BoringSSL… 🙈

There are traces of SIV in OpenSSL's documentation (e.g. [`EVP_CIPHER_CTX_set_padding`](https://www.openssl.org/docs/manmaster/man3/EVP_CIPHER_CTX_set_padding.html) mentions `EVP_CIPH_SIV_MODE`, but I haven't been able to find a dedicated API… however I've never used OpenSSL myself directly, so this doesn't mean much 😅

And while AES-GCM-SIV is faster (and a better choice if available), the performance penalty of the classic algorithms should nonetheless be negligible in most circumstances due to AESNI or NEON. I/O, compression and Borg's logic are probably much more of a bottleneck than 2*AES per chunk.

**Comment by ThomasWaldmann** [comment]
Would be cool if PR #6463 could get some review.
